ably/realtime_presence.go

```diff
@@ -100,23 +130,60 @@ func (pres *RealtimePresence) syncWait() {
 	pres.syncMtx.Unlock()
 }
 
-func syncSerial(msg *protocolMessage) string {
-	if i := strings.IndexRune(msg.ChannelSerial, ':'); i != -1 {
-		return msg.ChannelSerial[i+1:]
+// RTP18
+func syncSerial(msg *protocolMessage) (noChannelSerial bool, syncSequenceId string, syncCursor string) {
+	if empty(msg.ChannelSerial) { // RTP18c
+		noChannelSerial = true
+		return
+	}
+	// RTP18a
+	serials := strings.Split(msg.ChannelSerial, ":")
+	syncSequenceId = serials[0]
+	if len(serials) > 1 {
+		syncCursor = serials[1]
 	}
-	return ""
+	return false, syncSequenceId, syncCursor
 }
 
-func (pres *RealtimePresence) onAttach(msg *protocolMessage) {
-	serial := syncSerial(msg)
+func (pres *RealtimePresence) enterMembers(internalMembers []*PresenceMessage) {
+	for _, member := range internalMembers {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		// RTP17g
+		err := pres.EnterClient(ctx, member.ClientID, member.Data)
+		// RTP17e
+		if err != nil {
+			pres.channel.log().Errorf("Error for internal member presence enter with id %v, clientId %v, err %v", member.ID, member.ClientID, err)
+			pres.channel.emitErrorUpdate(newError(91004, err), true)
+		}
+		cancel()
+	}
+}
+
+func (pres *RealtimePresence) onAttach(msg *protocolMessage, isAttachWithoutMessageLoss bool) {
```

Comment:
The naming of `isAttachWithoutMessageLoss` seems misleading. That'll be set to true in any situation where you need to re-enter your members, which includes both situations where you do and don't have message loss

Reply:
Let's say, if you `attached` channel when previous state was `detached` or `attaching`, do we get `resume` as false in such a case?

Reply:
So, we should call it `isFreshAttach` wdyt?